File: core/src/main/java/ch/dissem/bitmessage/entity/valueobject/extended/Vote.java
package ch.dissem.bitmessage.entity.valueobject.extended;

import ch.dissem.bitmessage.entity.Plaintext;
import ch.dissem.bitmessage.entity.valueobject.ExtendedEncoding;
import ch.dissem.bitmessage.entity.valueobject.InventoryVector;
import ch.dissem.msgpack.types.*;

import java.io.IOException;
import java.util.Objects;

import static ch.dissem.bitmessage.utils.Strings.str;
import static ch.dissem.msgpack.types.Utils.mp;

/**
 * Extended encoding type 'vote'. Specification still outstanding, so this will need some work.
 */
public class Vote implements ExtendedEncoding.ExtendedType {
    private static final long serialVersionUID = -8427038604209964837L;

    public static final String TYPE = "vote";

    private InventoryVector msgId;
    private String vote;

    private Vote(Builder builder) {
        msgId = builder.msgId;
        vote = builder.vote;
    }

    @Override
    public String getType() {
        return TYPE;
    }

    public InventoryVector getMsgId() {
        return msgId;
    }

    public String getVote() {
        return vote;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Vote vote1 = (Vote) o;
        return Objects.equals(msgId, vote1.msgId) &&
            Objects.equals(vote, vote1.vote);
    }

    @Override
    public int hashCode() {
        return Objects.hash(msgId, vote);
    }

    @Override
    public MPMap<MPString, MPType<?>> pack() throws IOException {
        MPMap<MPString, MPType<?>> result = new MPMap<>();
        result.put(mp(""), mp(TYPE));
        result.put(mp("msgId"), mp(msgId.getHash()));
        result.put(mp("vote"), mp(vote));
        return result;
    }

    public static class Builder {
        private InventoryVector msgId;
        private String vote;

        public ExtendedEncoding up(Plaintext message) {
            msgId = message.getInventoryVector();
            vote = "1";
            return new ExtendedEncoding(new Vote(this));
        }

        public ExtendedEncoding down(Plaintext message) {
            msgId = message.getInventoryVector();
            vote = "1";
            return new ExtendedEncoding(new Vote(this));
        }

        public Builder msgId(InventoryVector iv) {
            this.msgId = iv;
            return this;
        }

        public Builder vote(String vote) {
            this.vote = vote;
            return this;
        }

        public ExtendedEncoding build() {
            return new ExtendedEncoding(new Vote(this));
        }
    }

    public static class Unpacker implements ExtendedEncoding.Unpacker<Vote> {
        @Override
        public String getType() {
            return TYPE;
        }

        @Override
        public Vote unpack(MPMap<MPString, MPType<?>> map) {
            Vote.Builder builder = new Vote.Builder();
            MPType<?> msgId = map.get(mp("msgId"));
            if (msgId instanceof MPBinary) {
                builder.msgId(new InventoryVector(((MPBinary) msgId).getValue()));
            }
            builder.vote(str(map.get(mp("vote"))));
            return new Vote(builder);
        }
    }
}

File: core/src/main/kotlin/ch/dissem/bitmessage/entity/valueobject/extended/Vote.kt
/*
 * Copyright 2017 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.entity.valueobject.extended

import ch.dissem.bitmessage.entity.Plaintext
import ch.dissem.bitmessage.entity.valueobject.ExtendedEncoding
import ch.dissem.bitmessage.entity.valueobject.InventoryVector
import ch.dissem.bitmessage.utils.Strings.str
import ch.dissem.msgpack.types.MPBinary
import ch.dissem.msgpack.types.MPMap
import ch.dissem.msgpack.types.MPString
import ch.dissem.msgpack.types.MPType
import ch.dissem.msgpack.types.Utils.mp

/**
 * Extended encoding type 'vote'. Specification still outstanding, so this will need some work.
 */
data class Vote constructor(val msgId: InventoryVector, val vote: String) : ExtendedEncoding.ExtendedType {

    override val type: String = TYPE

    override fun pack(): MPMap<MPString, MPType<*>> {
        val result = MPMap<MPString, MPType<*>>()
        result.put("".mp, TYPE.mp)
        result.put("msgId".mp, msgId.hash.mp)
        result.put("vote".mp, vote.mp)
        return result
    }

    class Builder {
        private var msgId: InventoryVector? = null
        private var vote: String? = null

        fun up(message: Plaintext): ExtendedEncoding {
            msgId = message.inventoryVector
            vote = "1"
            return ExtendedEncoding(Vote(msgId!!, vote!!))
        }

        fun down(message: Plaintext): ExtendedEncoding {
            msgId = message.inventoryVector
            vote = "-1"
            return ExtendedEncoding(Vote(msgId!!, vote!!))
        }

        fun msgId(iv: InventoryVector): Builder {
            this.msgId = iv
            return this
        }

        fun vote(vote: String): Builder {
            this.vote = vote
            return this
        }

        fun build(): ExtendedEncoding {
            return ExtendedEncoding(Vote(msgId!!, vote!!))
        }
    }

    class Unpacker : ExtendedEncoding.Unpacker<Vote> {
        override val type: String
            get() = TYPE

        override fun unpack(map: MPMap<MPString, MPType<*>>): Vote {
            val msgId = InventoryVector.fromHash((map["msgId".mp] as? MPBinary)?.value) ?: throw IllegalArgumentException("data doesn't contain proper msgId")
            val vote = str(map["vote".mp]) ?: throw IllegalArgumentException("no vote given")
            return Vote(msgId, vote)
        }
    }

    companion object {
        @JvmField
        val TYPE = "vote"
    }
}
